File: shared/src/commonMain/kotlin/com/emfmeter/util/UnitConverter.kt
package com.emfmeter.util

import com.emfmeter.domain.EMFUnit

/**
 * Utility for converting EMF values between units and formatting for display.
 */
object UnitConverter {

    /**
     * Convert a value from one unit to another.
     */
    fun convert(value: Float, from: EMFUnit, to: EMFUnit): Float {
        if (from == to) return value
        val microTesla = from.toMicroTesla(value)
        return to.fromMicroTesla(microTesla)
    }

    /**
     * Format a value for display with appropriate decimal places.
     */
    fun formatValue(value: Float, unit: EMFUnit): String {
        return when (unit) {
            EMFUnit.MICRO_TESLA -> formatFloat(value, 1)
            EMFUnit.MILLI_GAUSS -> formatFloat(value, 1)
            EMFUnit.GAUSS -> formatFloat(value, 3)
        }
    }

    /**
     * Format a float to a string with specified decimal places.
     * Uses platform-agnostic formatting.
     */
    private fun formatFloat(value: Float, decimals: Int): String {
        val factor = pow10(decimals)
        val rounded = kotlin.math.round(value * factor) / factor

        val intPart = rounded.toLong()
        val decPart = ((rounded - intPart) * factor).toLong()

        return if (decimals > 0) {
            "$intPart.${decPart.toString().padStart(decimals, '0')}"
        } else {
            intPart.toString()
        }
    }

    private fun pow10(n: Int): Float {
        var result = 1f
        repeat(n) { result *= 10f }
        return result
    }

    /**
     * Get the maximum display value for a unit (based on MAX_VALUE_UT = 200 µT).
     */
    fun getMaxDisplayValue(unit: EMFUnit): Float {
        return unit.fromMicroTesla(200f)
    }

    /**
     * Generate scale labels for the analog meter.
     */
    fun getScaleLabels(unit: EMFUnit, divisions: Int): List<String> {
        val maxValue = getMaxDisplayValue(unit)
        return (0..divisions).map { i ->
            val value = (maxValue * i / divisions)
            when (unit) {
                EMFUnit.GAUSS -> formatFloat(value, 1)
                else -> value.toInt().toString()
            }
        }
    }
}

File: ios/EMFMeter/Sources/Utils/UnitConverter.swift
import Foundation

/// Utility for converting EMF values between units and formatting for display.
enum UnitConverter {

    /// Convert a value from one unit to another.
    static func convert(_ value: Float, from: EMFUnit, to: EMFUnit) -> Float {
        if from == to { return value }
        let microTesla = from.toMicroTesla(value)
        return to.fromMicroTesla(microTesla)
    }

    /// Format a value for display with appropriate decimal places.
    static func formatValue(_ value: Float, unit: EMFUnit) -> String {
        switch unit {
        case .microTesla:
            return String(format: "%.1f", value)
        case .milliGauss:
            return String(format: "%.1f", value)
        case .gauss:
            return String(format: "%.3f", value)
        }
    }

    /// Get the maximum display value for a unit (based on MAX_VALUE_UT = 200 µT).
    static func getMaxDisplayValue(for unit: EMFUnit) -> Float {
        unit.fromMicroTesla(200)
    }

    /// Generate scale labels for the analog meter.
    static func getScaleLabels(for unit: EMFUnit, divisions: Int) -> [String] {
        let maxValue = getMaxDisplayValue(for: unit)
        return (0...divisions).map { i in
            let value = maxValue * Float(i) / Float(divisions)
            switch unit {
            case .gauss:
                return String(format: "%.1f", value)
            default:
                return String(Int(value))
            }
        }
    }
}
